arcgis-ios-sdk-samples/Cloud & Portal/Search for webmap by keyword/WebMapViewController.swift

```diff
@@ -17,33 +17,48 @@
 import UIKit
 import ArcGIS
 
-class WebMapViewController: UIViewController, AGSAuthenticationManagerDelegate {
+class WebMapViewController: UIViewController {
 
     @IBOutlet var mapView:AGSMapView!
     
-    var portalItem:AGSPortalItem!
-    var map:AGSMap!
+    var portalItem:AGSPortalItem?
     
     override func viewDidLoad() {
         super.viewDidLoad()
-
-        //initialize map with selected portal item
-        self.map = AGSMap(item: portalItem)
         
-        AGSAuthenticationManager.shared().delegate = self
+        guard let portalItem = portalItem else {
+            return
+        }
         
-        self.title = self.map.item?.title
+        // register to be notified about authentication challenges
+        AGSAuthenticationManager.shared().delegate = self
+
+        // initialize a map with the portal item
+        let map = AGSMap(item: portalItem)
+        // assign the map to the map view
+        mapView.map = map
         
-        //assign map to the map view
-        self.mapView.map = self.map
+        title = portalItem.title
     }
+}
 
-    
-    //MARK: - AGSAuthenticationManagerDelegate
+extension WebMapViewController: AGSAuthenticationManagerDelegate {
     
     func authenticationManager(_ authenticationManager: AGSAuthenticationManager, didReceive challenge: AGSAuthenticationChallenge) {
-        SVProgressHUD.showError(withStatus: "Web map access denied")
+        
+        // if a challenge is received, then the portal item is not fully public and cannot be displayed
+        
+        // stop attempts at map loading
+        mapView.map = nil
+        
+        // don't present this challenge
         challenge.cancel()
-        _ = self.navigationController?.popViewController(animated: true)
+        
+        // close this view controller
+        navigationController?.popViewController(animated: true)
```

Comment:
It would be best to get rid of this if at all possible. `WebMapViewController` shouldn't know that it is embedded in a navigation controller, nor should it know which view controller to pop to (the one immediately before it). It could either inform a delegate that a challenge was received, or have an authentication challenge handler, or the showing view controller could be the authentication manager's delegate and pop back to itself. Unfortunately there isn't a counterpart to `showDetailViewController(_:sender:)` for hiding view controllers without needing to know how they were shown.

It would be really nice if the showing view controller could determine whether authentication is needed before showing the web map view controller, but I don't think that is currently possible. =(